### ISSUE TITLE
Contrastive loss implementation

### ISSUE THREAD

**Issue by longyuxi** [computational-cell-analytics/micro-sam#769]
Hello,

Thanks for this great software package. It has been extremely easy to use and performs great.

For my specific application (light sheet microscopy), I deduced that it may be beneficial to use a contrastive loss to guide to model to consistently segment consecutive z-levels. Would you mind providing some ideas to how to implement this in this package?

I am looking at the trainer in `sam_trainer.py`, and currently it has a loss function like this:

```
 def _compute_loss(self, batched_outputs, y_one_hot):
        ...
        loss = mask_loss + iou_regression_loss

        return loss, mask_loss, iou_regression_loss
```

Would say, defining a new loss function differentiable through autograd, say `contrastive_loss()`, and modifying the above function to be like

```
 def _compute_loss(self, batched_outputs, y_one_hot):
        ...
        c_loss = self.contrastive_loss(self)
        loss = mask_loss + iou_regression_loss + c_loss


        return loss, mask_loss, iou_regression_loss
```

hypothetically work?

**Comment by constantinpape** [comment]
Yes, your idea should in principle work. However, if you want to apply a contrastive loss across slices in a z-stack this will require some more changes. In particular, you will have to update the training so that it operates on z-stacks. This is currently not the case, as each image in the data will be treated independently. So for a 3D training dataset each slice will be treated as an individual image.

**Comment by constantinpape** [comment]
Closing this due to inactivity. Feel free to reopen if you have further questions.
